Added signed header verification

Manav-Aggarwal opened the pull request:
<!--
Please read and fill out this form before submitting your PR.

Please make sure you have reviewed our contributors guide before submitting your
first PR.
-->

## Overview
Resolves: #1027, Resolves: #1048
<!-- 
Please provide an explanation of the PR, including the appropriate context,
background, goal, and rationale. If there is an issue with this information,
please provide a tl;dr and link the issue. 
-->

## Checklist

<!-- 
Please complete the checklist to ensure that the PR is ready to be reviewed.

IMPORTANT:
PRs should be left in Draft until the below checklist is completed.
-->

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

MSevey commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+		return errors.New("Last header hash does not match hash of previous header")
+	}
+	if !bytes.Equal(untrstH.LastCommitHash[:], sH.getLastCommitHash()) {
+		return errors.New("Last header hash does not match hash of previous header")
```
copy paste. i assume the error message should reference commit hash?

Manav-Aggarwal replied:
fixed

gupadhyaya commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+
+func (sH *SignedHeader) getLastCommitHash() []byte {
+	lastABCICommit := tmtypes.Commit{
+		Height: int64(sH.BaseHeader.Height),
```
use `sH.Height()`

gupadhyaya commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+
+func (sH *SignedHeader) getLastCommitHash() []byte {
+	lastABCICommit := tmtypes.Commit{
+		Height: int64(sH.Height()),
```
`sH.Height()` is already `int64`. no need for additional typecast.

nashqueue commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+		return err
+	}
+	if !bytes.Equal(untrstH.LastHeaderHash[:], sH.Header.Hash()) {
+		return errors.New("Last header hash does not match hash of previous header")
```
We should use the go-header Error struct for verification here https://github.com/celestiaorg/go-header/blob/main/errors.go . Also print out the hashes that are not matching for more complete error messages. If this is applicable, please do it for the other errors as well.

nashqueue commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
 	}
+	err := untrstH.ValidateBasic()
+	if err != nil {
+		return err
```
You can wrap the error in the VerifyError struct here as well. Please take a look at the comment below.

nashqueue commented on types/signed_header.go:
```diff
@@ -22,7 +26,41 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
 		// and is a programmer bug
 		panic(fmt.Errorf("%T is not of type %T", untrst, untrstH))
 	}
+	err := untrstH.ValidateBasic()
```
Pulling the verification steps from ValidateBasic() from 'validation.go' into here might make sense so we can verify a type in one place. I suggest we start splitting the types package into its packages. Let's say a header package with dedicated tests for verification of the header, for example. thoughts cc @MSevey

MSevey replied:
If all the code related to headers fits nicely into a single file then I don't think we need to sub divide the packages.  

One types folder with a single file and test file per category is reasonable imo. 
If we we need to start expanding into multiple files to handle all the types code associated with the header for example, then I think it would make more sense to sub divide the package so that you were just pre-pending `header_*.go` to everything.

nashqueue replied:
Interesting I suggest we revisit this when we think about a more general refactor for cleaner code.  What's your opinion on my first point ?
> Pulling the verification steps from ValidateBasic() from 'validation.go' into here might make sense so we can verify a type in one place.

MSevey replied:
> pulling the verification steps from ValidateBasic() from 'validation.go' into here might make sense so we can verify a type in one place.

Yes agree, we should think of files as scope barriers. So `signed_header.go` should contain all the code related to signed headers.

MSevey replied:
This honestly can be applied to the rest of the code in the `validation.go` file.

Manav-Aggarwal replied:
Removed `validation.go` and moved all related validation logic to respective files.

nashqueue commented:
Another thing that we should consider is that we have 2 options for verifying a header. Either if it is adjacent, we check it through the last hash etc. OR if it is in the future, we can assume it is correct because the Sequencer signed it and can be fraud proven later.

Manav-Aggarwal commented:
> Another thing that we should consider is that we have 2 options for verifying a header. Either if it is adjacent, we check it through the last hash etc. OR if it is in the future, we can assume it is correct because the Sequencer signed it and can be fraud proven later.

We should discuss this and track it in an issue.

nashqueue reviewed (approved):
LGTM Last point in a non-blocking refactor and can be done in a follow up

MSevey commented on types/signed_header.go:
```diff
@@ -22,7 +25,49 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+		lastABCICommit.Signatures = append(lastABCICommit.Signatures, commitSig)
+	}
+
+	if len(sH.Commit.Signatures) == 1 {
```
can we add a comment to this? It's not obvious what we are doing here and why it only applies to len of 1

MSevey commented on types/signed_header.go:
```diff
@@ -22,7 +25,49 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
 var _ header.Header = &SignedHeader{}
+
+func (sH *SignedHeader) getLastCommitHash() []byte {
+	lastABCICommit := tmtypes.Commit{
+		Height: sH.Height(),
+		Round:  0,
+		BlockID: tmtypes.BlockID{
+			Hash:          tmbytes.HexBytes(sH.Hash()),
+			PartSetHeader: tmtypes.PartSetHeader{},
+		},
+	}
+	for _, sig := range sH.Commit.Signatures {
+		commitSig := tmtypes.CommitSig{
+			BlockIDFlag: tmtypes.BlockIDFlagCommit,
+			Signature:   sig,
+		}
+		lastABCICommit.Signatures = append(lastABCICommit.Signatures, commitSig)
+	}
```
Whenever you use `append` you are reallocating memory.  To avoid this, you can pre-allocate the capacity of the slice.
```suggestion
	lastABCICommit := tmtypes.Commit{
		Height: sH.Height(),
		Round:  0,
		BlockID: tmtypes.BlockID{
			Hash:          tmbytes.HexBytes(sH.Hash()),
			PartSetHeader: tmtypes.PartSetHeader{},
		},
		Signatures: make([]tmtypes.CommitSig,0,len(sH.Commit.Signatures)
	}
	for _, sig := range sH.Commit.Signatures {
		commitSig := tmtypes.CommitSig{
			BlockIDFlag: tmtypes.BlockIDFlagCommit,
			Signature:   sig,
		}
		lastABCICommit.Signatures = append(lastABCICommit.Signatures, commitSig)
	}
```

Manav-Aggarwal replied:
I don't think you should append when you've already made allocated. Needs to use `lastABCICommit[i]` most likely.

Manav-Aggarwal replied:
This `getLastCommitHash` is taken from https://github.com/rollkit/rollkit/blob/7bd5df91ed0f48d72c4caac7de069b48b6c1f608/state/executor.go#L478
and
https://github.com/rollkit/rollkit/blob/e4f8bd441b49af44de04b637031520eab437011f/conv/abci/block.go#L119

I tried using the existing method but there's a circular dependency if I try to do that :/

nashqueue replied:
That's an interesting problem. If we extracted the header logic into its own file, we could import logic built on other types without having a circular dependency. For now, I think that is fine but we should clean this up again in a follow-up PR.

MSevey reviewed (changes requested):
some minor edits, but then will be good

tzdybal commented on types/signed_header.go:
```diff
@@ -22,7 +25,70 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+		return err
+	}
+
+	// Handle Based Rollup case
+	if h.Validators == nil || len(h.Validators.Validators) == 0 {
+		return nil
+	}
```
I really don't like those lines. It's exiting fast, and IMHO is very error prone. In general, I would like to see something like:

```go
switch h.rollupType() {
	case based:
		// validation for based rollup
	case optimistic:
		// validation for optimistic rollup
	// ...
}

tzdybal replied:
For example, if `len(h.Validators.Validators) == 0` signature validation is completely skipped, and it is not clear if it's a bug or a feature.

Manav-Aggarwal replied:
Yes, there is no signature verification in based rollup since there is no specified block proposer. Also, note that this code is simply moved from `validation.go` to `signed header.go` because of the suggestion here: https://github.com/rollkit/rollkit/pull/1026#discussion_r1238280719.

Created an issue for your suggestion: https://github.com/rollkit/rollkit/issues/1042 👍

tzdybal commented on types/signed_header.go:
```diff
@@ -22,7 +25,70 @@ func (sH *SignedHeader) Verify(untrst header.Header) error {
+	}
+
+	signature := h.Commit.Signatures[0]
+	proposer := h.Validators.GetProposer()
```
`h.Validators.GetProposer()` can return `nil`, and this will cause a panic.

Manav-Aggarwal replied:
It technically can never return `nil` because of the check above that returns if `len(h.Validators.Validators) == 0`. Another reason to modify the `ValidateBasic` code to make it more clearer which https://github.com/rollkit/rollkit/issues/1042 should handle.

tzdybal reviewed (changes requested):
`New and updated code has new and/or updated testing` - this is false. Code is not covered by unit tests at all. I see that there is a subtask in https://github.com/rollkit/rollkit/issues/1031, but IMO creating unit tests should not be separated from implementation.

Manav-Aggarwal commented on types/block.go:
```diff
@@ -61,3 +63,66 @@ type Signature []byte
+	if len(c.Signatures) == 1 {
+		lastABCICommit.Signatures[0].ValidatorAddress = proposerAddress
+		lastABCICommit.Signatures[0].Timestamp = header.Time()
+	}
```
This mimics the behavior in the executorhttps://github.com/rollkit/rollkit/blob/4c951e48862cdd00e04730336f0c0c9fd00c1101/state/executor.go#L480C38-L482

What happens when multiple signatures are present here? Is there a reason it doesn't return an error here?

MSevey replied:
Agree that the comment makes me expect an error. 

Recommend follow up discussion. We can covert this to an issue and apply any updates in both places in a separate PR.

Manav-Aggarwal replied:
Created an issue here: https://github.com/rollkit/rollkit/issues/1050

Manav-Aggarwal commented:
> `New and updated code has new and/or updated testing` - this is false. Code is not covered by unit tests at all. I see that there is a subtask in #1031, but IMO creating unit tests should not be separated from implementation.

Added header verification tests as well.

MSevey commented on types/signed_header_test.go:
```diff
@@ -0,0 +1,104 @@
+func TestVerify(t *testing.T) {
+	trusted, privKey, err := GetRandomSignedHeader()
+	require.NoError(t, err)
+	time.Sleep(time.Second)
```
😢  why you do me dirty like this and through in bare sleeps in tests haha

What's the need for the sleep? 

If this is a copy paste, let's create a follow up issue to remove so we don't slow down this PR with debugging a potentially flaky test.

Manav-Aggarwal replied:
Sorry, this seemed to be the most straightforward way, but in this case, the sleep is to advance the system time logic so that the next random signed header has a later time (see [code ref](https://github.com/rollkit/rollkit/blob/c7e9725e0fb3bb23f312dad780fa0066333f11d9/types/test_utils.go#L73)), since a verification check in the signed header uses `time.Now()`, modifying the time manually for the new header leads to the error: code [ref.](https://github.com/rollkit/rollkit/blob/c7e9725e0fb3bb23f312dad780fa0066333f11d9/types/header.go#L140C22-L140C59) Anything less than a second leads to running into this error: [code ref](https://github.com/rollkit/rollkit/blob/c7e9725e0fb3bb23f312dad780fa0066333f11d9/types/header.go#L133) since `After` truncates the time to be in seconds.

MSevey commented on types/signed_header_test.go:
```diff
@@ -0,0 +1,104 @@
+	require.NoError(t, err)
+	tests := []struct {
+		prepare func() *SignedHeader
+		err     bool
```
[Follow Up]

we should have the expected error here. Then in the test execution we are validating that we are erroring exactly as expected and are not erroring for some other reason. This will help better future proof the code if there are functions that have a chain of checks and we re-order them.

Manav-Aggarwal replied:
Created issue here: https://github.com/rollkit/rollkit/issues/1049

MSevey reviewed (approved):
LGTM, all my comments can be follow ups.

nashqueue reviewed (approved):
Thank you! LGTM, You can resolve conversations when followups are created

tzdybal reviewed (approved):
LGTM

Manav-Aggarwal commented:
Removed when enforcing centralized sequencer.
